oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java

```diff
@@ -62,77 +61,95 @@
       Server updateServer = null;
+      PasswordGenerator generate = new PasswordGenerator();
+      Types.ServerType serverType = Types.ServerType.cloud;
 
       final String loginUser = isNullOrEmpty(options.getLoginUser()) ? "root" : options.getLoginUser();
-      final String password = options.hasLoginPassword() ? options.getLoginPassword() : passwordGenerator.generate();
+      final String password = options.hasLoginPassword() ? options.getLoginPassword() : generate.generate();
       final String privateKey = options.hasLoginPrivateKey() ? options.getPrivateKey() : null;
       final org.jclouds.compute.domain.Image image = template.getImage();
       final int[] inboundPorts = template.getOptions().getInboundPorts();
-
-      //prepare hdds to provision
-      List<? extends Volume> volumes = hardware.getVolumes();
-      List<Hdd.CreateHdd> hdds = new ArrayList<Hdd.CreateHdd>();
-
-      for (final Volume volume : volumes) {
-         try {
-            //check if the bootable device has enough size to run the appliance(image).
-            float minHddSize = volume.getSize();
-            if (volume.isBootDevice()) {
-               SingleServerAppliance appliance = api.serverApplianceApi().get(image.getId());
-               if (appliance.minHddSize() > volume.getSize()) {
-                  minHddSize = appliance.minHddSize();
-               }
-            }
-            Hdd.CreateHdd hdd = Hdd.CreateHdd.create(minHddSize, volume.isBootDevice());
-            hdds.add(hdd);
-         } catch (Exception ex) {
-            throw Throwables.propagate(ex);
-
+      String imageId = image.getId();
+      Hardware hardwareModel = generateHardwareRequest.isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage = null;
+
+      //choose the correct image based on the server type baremetal or cloud
+      if (hardwareModel != null) {
+         if (isBaremetal) {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.baremetal);
+         } else {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         }
+         if (workingImage != null) {
+            imageId = workingImage.id();
+         }
+      } else {
+         //check if image is of Type IMAGE
+         workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         if (workingImage != null) {
+            imageId = workingImage.id();
          }
       }
 
-      // provision server
-      Server server = null;
-      Double cores = ComputeServiceUtils.getCores(hardware);
-      Double ram = (double) hardware.getRam();
-      if (ram < 1024) {
-         ram = 0.5;
-      } else {
-         ram = ram / 1024;
+      //configuring Firewall rules
+      Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
```

Comment:
Is this change strictly related to bare metal support? seems to me that this PR is trying to change too many things, would it be possible for you to split up into multiple smaller PR so to have a better/easier review?

Reply:
This has just been moved to the GenerateHardwareRequest.java to organize code just as you request to separate the creation of the hardware request.